Source/Core/DiscIO/FileSystemGCWii.cpp

```diff
@@ -19,83 +21,309 @@
+
+	// read the whole FST
+	u64 fst_offset = static_cast<u64>(m_rVolume->Read32(0x424, m_Wii)) << GetOffsetShift();
+	u32 raw_fst_size = m_rVolume->Read32(0x428, m_Wii);
+	u64 fst_size = static_cast<u64>(m_rVolume->Read32(0x428, m_Wii)) << GetOffsetShift();
+	if (fst_size < 0xC)
+		return;
+	m_file_system_table.resize(fst_size + 1);
+	if (!m_rVolume->Read(fst_offset, fst_size, m_file_system_table.data(), m_Wii))
+		return;
+
+	// Just in case the last string in the name table isn't null terminated
+	m_file_system_table[fst_size - 1] = 0;
+
+	// Create the root object
+	m_root = CFileInfoGCWii(m_file_system_table.data(), m_Wii, raw_fst_size);
+
+	if (!m_root.IsDirectory())
+		return;
+
+	// If we haven't returned yet, everything succeeded
+	m_Valid = true;
 }
 
 CFileSystemGCWii::~CFileSystemGCWii()
 {
-	m_FileInfoVector.clear();
 }
 
-u64 CFileSystemGCWii::GetFileSize(const std::string& _rFullPath)
+const IFileInfo& CFileSystemGCWii::GetRoot() const
 {
-	if (!m_Initialized)
-		InitFileSystem();
-
-	const SFileInfo* pFileInfo = FindFileInfo(_rFullPath);
+	return m_root;
+}
 
-	if (pFileInfo != nullptr && !pFileInfo->IsDirectory())
-		return pFileInfo->m_FileSize;
+std::unique_ptr<IFileInfo> CFileSystemGCWii::FindFileInfo(const std::string& path) const
+{
+	if (!IsValid())
+		return nullptr;
 
-	return 0;
+	return FindFileInfo(path, m_root);
 }
 
-const std::string CFileSystemGCWii::GetFileName(u64 _Address)
+std::unique_ptr<IFileInfo> CFileSystemGCWii::FindFileInfo(const std::string& path, const IFileInfo& file_info) const
 {
-	if (!m_Initialized)
-		InitFileSystem();
+	// Given a path like "directory1/directory2/fileA.bin", this method will
+	// find directory1 and then call itself to search for "directory2/fileA.bin".
+
+	// If there's nothing left to search for, return the current entry
+	if (path.empty())
+		return std::unique_ptr<IFileInfo>(new CFileInfoGCWii((const CFileInfoGCWii&)file_info));
 
-	for (auto& fileInfo : m_FileInfoVector)
+	// It's only possible to search in directories. Searching in a file is an error
+	if (!file_info.IsDirectory())
+		return nullptr;
+
+	std::string searching_for;
+	std::string rest_of_path;
+	size_t first_dir_separator = path.find('/');
+	if (first_dir_separator != std::string::npos)
+	{
+		searching_for = path.substr(0, first_dir_separator);
+		rest_of_path = path.substr(first_dir_separator + 1);
+	}
+	else
+	{
+		searching_for = path;
+		rest_of_path = "";
+	}
+
+	for (const IFileInfo& child : file_info)
 	{
-		if ((fileInfo.m_Offset <= _Address) &&
-		    ((fileInfo.m_Offset + fileInfo.m_FileSize) > _Address))
+		if (child.GetName() == searching_for)
 		{
-			return fileInfo.m_FullPath;
+			// A match is found. The rest of the path is passed on to finish the search
+			std::unique_ptr<IFileInfo> result = FindFileInfo(rest_of_path, child);
+			if (result)
+				return result;
 		}
 	}
 
-	return "";
+	return nullptr;
+}
+
+std::unique_ptr<IFileInfo> CFileSystemGCWii::FindFileInfo(u64 disc_offset) const
```

Comment:
The style guide says "Try to avoid using raw pointers (pointers allocated with `new`) as much as possible", and I wouldn't want any of the callers to forget deleting the pointer. Callers wanting to use a different type of ownership is a valid point, but I don't want to return a raw pointer unless there's consensus for doing so. (cc @Lioncash?)